### ISSUE TITLE
`kDebugMode` doesn't respect `assert` function context

### ISSUE THREAD

**Issue by maRci002** [dart-lang/linter#3878]
**Describe the issue**
```dart
void main() {
  final array = <Object?>[];

  assert(() {
    final contains = array.contains(null);
    // ... other calculations

    print('Array contains null: $contains'); // Avoid `print` calls in production code.

    // ... other things

    return true;
  }());
}
```

Should I use just wrap my code in `kDebugMode` block?

**Question**
If I were using `assert` without function call then it's condition is evaluated only in `debug` mode?

```dart
void main() {
  final result = <Object?>[];

  assert(result.contains(null));
}
```

Or `result.contains(null)` will be evaluated in `release` / `profile` mode since it might have side effects:

```dart
void main() {
  final result = <Object?>[];

  final condition = result.contains(null);
  assert(condition);
}
```

**Comment by goderbauer** [comment]
From the issue it is actually not clear to me what the question is?

**Comment by bwilkerson** [comment]
Asserts are compiled out unless they have been enabled (typically only in debug mode), so the expression will not be evaluated when they are not enabled.

**Comment by maRci002** [comment]
I just run my second question's code in `release` and compiler was smart enough to remove `functionWithSideEffect` call.

I thought I have to wrap it in an anonymous function like this:
```dart
  assert(
    () {
      return functionWithSideEffect();
    }(),
    'Is this message calculated in release mode? $a',
  );
```

**Comment by lrhn** [comment]
If this lint is only recommended for Flutter code, then it should be documented as such, as "Don't use this lint for non-Flutter projects".

If it's intended for general use, but with extra recognition of Flutter specific features, then it should probably also consider `assert` statement as not being production code.
(Detecting functions only called from `assert`s is slightly harder, but inline code should be easy.)
